[M] 바우처를 생성할 수 있다 (#14)

KwonYeKyeong opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->

## 👩‍💻 요구 사항과 구현 내용 
- 중복된 바우처는 생성할 수 없다
- 음의 값을 가지는 바우처는 생성할 수 없다
- 한계 값을 초과한 바우처는 생성할 수 없다

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->

hanjo8813 committed:
feat : user 도메인 생성자 추가

KwonYeKyeong committed:
[M] fix : voucher PK 생성 전략 변경

KwonYeKyeong committed:
[M] feat : 바우처를 생성할 수 있다 (#14)

KwonYeKyeong committed:
yaml 설정 파일 변경 (#27)

* [C] feat : 회원가입 기능 구현 (#21)

* chore : yaml 설정 변경

KwonYeKyeong committed:
fix : amazonaws 관련 yaml 설정 추가

KwonYeKyeong committed:
[M] feat : 중복된 바우처는 생성할 수 없다 (#14)

KwonYeKyeong committed:
[M] test : controller 테스트 코드 작성 (#14)

KwonYeKyeong committed:
Merge branch 'develop' into m/voucher

hanjo8813 commented on ourhome-market/src/main/java/com/armand/ourhome/market/voucher/dto/VoucherDto.java:
```diff
@@ -0,0 +1,28 @@
+import lombok.Setter;
+
+@NoArgsConstructor
+@Setter
```
`@Setter` 빼기

KwonYeKyeong replied:
Setter 지우세요!

KwonYeKyeong commented on ourhome-market/src/main/java/com/armand/ourhome/market/voucher/domain/FixedVoucher.java:
```diff
@@ -7,14 +7,15 @@
 public class FixedVoucher extends Voucher {
 
-  @Column(nullable = false)
+  @Column // (nullable = false)
```
주석 지우세요!!

KwonYeKyeong commented on ourhome-market/src/main/java/com/armand/ourhome/market/voucher/domain/PercentVoucher.java:
```diff
@@ -7,14 +7,15 @@
 public class PercentVoucher extends Voucher {
 
-  @Column(nullable = false)
+  @Column // (nullable = false)
```
주석 지우세요!!

KwonYeKyeong reviewed (commented):
지울게 많네요(머쓱)

seung-hun-h commented on ourhome-market/src/main/java/com/armand/ourhome/market/voucher/converter/VoucherConverter.java:
```diff
@@ -0,0 +1,47 @@
+    }
+    throw new IllegalArgumentException(
+        MessageFormat.format("잘못된 바우처 타입입니다. request.voucherType : {0}", request.getVoucherType()));
+  }
```
if문 분기보다는 enum type을 활용해서 인스턴스를 생성하는 편이 좋을 것 같습니다!

seung-hun-h commented on ourhome-market/src/main/java/com/armand/ourhome/market/voucher/dto/request/RequestVoucher.java:
```diff
@@ -10,20 +11,23 @@
 @NoArgsConstructor
 @Setter
```
@Setter가 필요 없을 것 같습니다

seung-hun-h commented on ourhome-market/src/main/java/com/armand/ourhome/market/voucher/repository/VoucherRepository.java:
```diff
@@ -1,8 +1,19 @@
 public interface VoucherRepository<T extends Voucher> extends JpaRepository<T, Long> {
 
+  @Query("select v from PercentVoucher v where v.percent = :percent and v.minLimit = :minLimit")
+  Optional<Voucher> findByPercentAndMinLimit(@Param("percent") Integer percent,
```
exists~~를 사용하면 boolean 값으로 반환 받을 수 있습니다!

seung-hun-h commented on ourhome-market/src/main/java/com/armand/ourhome/market/voucher/repository/VoucherRepository.java:
```diff
@@ -1,8 +1,19 @@
+  Optional<Voucher> findByPercentAndMinLimit(@Param("percent") Integer percent,
+      @Param("minLimit") Integer minLimit);
+
+  @Query("select v from FixedVoucher v where v.amount = :amount and v.minLimit = :minLimit")
+  Optional<Voucher> findByAmountAndMinLimit(@Param("amount") Integer amount,
```
쿼리 메소드나 JPQL을 직접 작성하는 것 둘 중에 하나면 하시면 좋을 것 같아요

seung-hun-h commented on ourhome-market/src/main/java/com/armand/ourhome/market/voucher/service/VoucherService.java:
```diff
@@ -1,5 +1,41 @@
+    if (saved.isPresent()) {
+      throw new BusinessException("이미 중복된 바우처가 있습니다", ErrorCode.INVALID_INPUT_VALUE);
+    }
+
```
DuplicateEntityException 처럼 의미 있는 예외 클래스를 만드는 것도 좋아보입니다!

seung-hun-h reviewed (commented):
예경님 코멘트 남겼습니다.

확인해주세요!

KwonYeKyeong committed:
[M] refactor : 코드리뷰 반영 (#34)

- RequestVoucher, VoucherDto @Setter 제거
- FixedVoucher, PercentVoucher 주석 제거
- DuplicateVoucherException 구체적인 예외 클래스 생성 및 반영

KwonYeKyeong committed:
[M] refactor : 코드리뷰 반영 (#34)

- enum type을 활용해서 인스턴스를 생성

seung-hun-h commented on ourhome-market/src/main/java/com/armand/ourhome/market/voucher/dto/VoucherType.java:
```diff
@@ -0,0 +1,20 @@
+      return FixedVoucher.of(value, minLimit);
+    }
+  },
+  PERCENT {
+    public FixedVoucher of(int value, int minLimit) {
+      return FixedVoucher.of(value, minLimit);
+    }
+  };
```
PercentVoucher가 생성되도록 변경해야할 것 같습니다
